zerver/tests/test_converters.py

```diff
@@ -0,0 +1,226 @@
+from decimal import Decimal
+from enum import Enum
+from typing import Dict, List, Literal, Optional, Sequence, Type, TypedDict
+
+import orjson
+from attr import define
+from cattrs import StructureHandlerNotFoundError, transform_error
+
+from zerver.lib.converters import (
+    MAX_INT_SIZE,
+    Color,
+    NonNegativeInt,
+    URLStr,
+    forbid_extra_keys,
+    format_exception,
+    json_converter,
+    register_hook_for_converters,
+    zulip_converter,
+)
+from zerver.lib.test_classes import ZulipTestCase
+
+
+class ConvertersTestCase(ZulipTestCase):
+    class _NoOutput:
+        def __repr__(self) -> str:
+            return "<NoOutput>"
+
+    NoOutput = _NoOutput()
+
+    def _assert_converter_results(
+        self,
+        input: object,
+        target_type: Type[object],
+        expected_messages: Sequence[str] = [],
+        use_json: bool = True,
+    ) -> object:
+        try:
+            if use_json:
+                output = json_converter.loads(orjson.dumps(input).decode(), target_type)
+            else:
+                output = zulip_converter.structure(input, target_type)
+        except BaseException as exc:
+            error_messages = transform_error(exc, path="input", format_exception=format_exception)
+            self.assertCountEqual(error_messages, expected_messages)
+            return self.NoOutput
+        else:
+            self.assertEqual(expected_messages, [], "Expected errors but none were raised")
+            return output
+
+    def verify_formatter_output(
+        self,
+        input: object,
+        target_type: Type[object],
+        *,
+        converter: Literal["json", "regular", "both"] = "both",
+        errors: Sequence[str] = [],
+        output: object = NoOutput,
+    ) -> None:
+        if converter in ("json", "both"):
```

Comment:
This is a another example of where I would try to avoid conditionals in tests.  You basically could have `verify_json_formatter_output` and `verify_regular_formatter_output`, and if you need to verify both, just call both.

Reply:
Yeah, I split it into `verify_formatter_output_json` and `verify_formatter_output_regular`.